### PR TITLE
xerces sandbox, c++11 cleanup initial version

### PR DESCRIPTION
**Pull request by tst-basilhuffman**
Added:

* Memory deallocation when STL containers and Qstrings emptied
* Method for reinitializing xerces, will look for places to test
* C++ style casting
* Range looping in places for readability, will revert of compatibility is still meeded

### CODE REVIEW

File: earth_enterprise/src/common/khxml/khdom.h [hunk 1]
```diff
@@ -281,10 +281,10 @@ ToElementWithChildName(khxml::DOMElement *elem,
                        const std::string &childTagName,
                        const std::vector<T> &vec)
 {
-  for (typename std::vector<T>::const_iterator iter = vec.begin();
-       iter != vec.end();
-       ++iter) {
-    AddElement(elem, childTagName, *iter);
+
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khdom.h]: Minor: looks like this function could be combined with the following 3 functions using a template:
```
template <class T>
void
ToElementWithChildName(khxml::DOMElement *elem,
                       const std::string &childTagName,
                       const T &container)
{

  for (const auto& iter : container)
  {
    AddElement(elem, childTagName, iter);
  }
}
```
That's not really the point of the ticket but it's worth doing while we're cleaning things up.

Comment (tst-lsavoie) [reply earth_enterprise/src/common/khxml/khdom.h]: There are probably other similar functions that can be combined as well.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 2]
```diff
@@ -40,30 +41,148 @@ ListElementTagName(const std::string &tagname)
...
+        }
+       notify(NFY_NOTICE,"successful xerces reinit...");
+    }
+
 };
 
 static khMutexBase xmlLibLock = KH_MUTEX_BASE_INITIALIZER;
 
-void InitializeXMLLibrary(void) throw()
+class XercesReInitGuard
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: We'll want to replace this with a standard mutex. I think `khLockGuard` would work fine.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 3]
```diff
@@ -40,50 +41,147 @@ ListElementTagName(const std::string &tagname)
...
+    khLockGuard outer(reInitOuterLock);
+    khLockGuard inner(reInitInnerLock);
+    InitializeXMLLibrary(true);
 }
 
 
 DOMDocument *
 CreateEmptyDocument(const std::string &rootTagname) throw()
 {
+  //khLockGuard outer(reInitOuterLock);
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: We can delete this unused code.

Comment (tst-lsavoie) [reply earth_enterprise/src/common/khxml/khxml.cpp]: Same comment several other places - we should delete the commented-out code.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 4]
```diff
@@ -40,50 +41,147 @@ ListElementTagName(const std::string &tagname)
 // This is used only in the following function
 class UsingXMLGuard
 {
-  friend void InitializeXMLLibrary(void) throw();
-
-  UsingXMLGuard(void) throw() {
-    try {
-      XMLPlatformUtils::Initialize();
+  friend void InitializeXMLLibrary(bool) throw();
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: I can see the value in keeping this code handy for further testing, but I don't think we want it in the master branch. Can we move this off to a separate WIP PR or something like that?

Comment (tst-lsavoie) [reply earth_enterprise/src/common/khxml/khxml.cpp]: The same goes for all of the changes to initializing the XML library.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 5]
```diff
@@ -40,50 +41,147 @@ ListElementTagName(const std::string &tagname)
...
 CreateEmptyDocument(const std::string &rootTagname) throw()
 {
+  //khLockGuard outer(reInitOuterLock);
+  //{
+  //khLockGuard inner(reInitInnerLock);
   InitializeXMLLibrary();
 
   try {
     DOMImplementation* impl =
-      DOMImplementationRegistry::getDOMImplementation(0);
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: I'd like to see the extra whitespace retained here. It's helpful to have the second line indented so that we know it's a continuation of the previous line. Same thing a few lines down.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 6]
```diff
@@ -196,44 +300,45 @@ WriteDocumentToString(DOMDocument *doc, std::string &buf) throw()
...
+        notify(NFY_WARN, "Unable to create DOM writer: %s",
+               XMLString::transcode(toCatch.msg));
+      } catch (...) {
+        notify(NFY_WARN, "Unable to create DOM writer: Unknown exception");
+      }
+  //}
   return success;
 }
 
+// eventually lock parser from creation to destruction
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: I don't understand this comment. It sounds like a bad idea to hold a lock from the time the parser is created to the time it is destroyed.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 7]
```diff
@@ -274,13 +379,18 @@ CreateDOMParser(void) throw()
...
+  DOMDocument* doc = nullptr;
+  //{
+  //khLockGuard inner(reInitInnerLock);
 
   try {
     // Note: parseURI doesn't handle missing files nicely...returns
     // invalid doc object. Must check file existence ourselves.
     if (khExists(filename)) {
-      doc = parser->parseURI(filename.c_str());
+        notify(NFY_NOTICE,"ReadDocument, filename: %s",
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: Is this meant to be temporary? It looks like debugging code.

Comment (tst-basilhuffman) [reply earth_enterprise/src/common/khxml/khxml.cpp]: Yes, will remove

File: earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp [hunk 8]
```diff
@@ -262,11 +268,14 @@ khResourceManager::EraseProvider(khResourceProviderProxy *proxy)
...
-    volumes.erase(*vn);
-  }
+  //for (std::vector<std::string>::const_iterator vn = volnames.begin();
+  //     vn != volnames.end(); ++vn) {
+  for (const auto& vn : volnames) {
+    delete volumes[vn];
+    volumes.erase(vn);
+  }
+  if (volumes.empty())
+      volumes.clear();
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: Minor: I like putting brackets around all if statements. The only time I make an exception is when I put the entire statement on one line: `if (volumes.empty()) volumes.clear()`

Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: Will do

File: earth_enterprise/src/common/khxml/khxml.cpp (hunk 5 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/common/khxml/khxml.cpp]: Will do

File: earth_enterprise/src/common/khxml/khxml.cpp (hunk 4 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/common/khxml/khxml.cpp]: Sure thing

File: earth_enterprise/src/common/khxml/khxml.cpp (hunk 6 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/common/khxml/khxml.cpp]: I guess this gets down to implementation. It doesn't really pertain to this PR and can be discussed with the next PR

File: earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp [hunk 9]
```diff
@@ -416,6 +423,13 @@ khResourceManager::TryActivate(void) throw()
       }
     }
 
+    for (const auto& p : providers)
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: Is this meant to replace the loop above? If so, we need to remove the original version.

File: earth_enterprise/src/common/khxml/khxml.h [hunk 10]
```diff
@@ -23,6 +23,8 @@
 #include <qstring.h>
 
 namespace khxml = xercesc;
+// forward declaration
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.h]: I think all the changes in this file can be reverted. It looks like they all have to do with the experimental method of reinitializing xerces, which we're not keeping in the code right now.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 11]
```diff
@@ -196,41 +262,41 @@ WriteDocumentToString(DOMDocument *doc, std::string &buf) throw()
...
+      } catch (const XMLException& toCatch) {
+        notify(NFY_WARN, "Unable to create DOM Writer: %s",
+               XMLString::transcode(toCatch.getMessage()));
+      } catch (const DOMException& toCatch) {
+        notify(NFY_WARN, "Unable to create DOM writer: %s",
+               XMLString::transcode(toCatch.msg));
+      } catch (...) {
+        notify(NFY_WARN, "Unable to create DOM writer: Unknown exception");
+      }
+  //}
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: Minor: remove commented out line.

File: earth_enterprise/src/common/khxml/khxml.cpp [hunk 12]
```diff
@@ -40,30 +41,96 @@ ListElementTagName(const std::string &tagname)
...
+//  times, and then follow with XMLPlatformUtils::Terminate() the same
+//  number of times, only the first XMLPlatformUtils::Initialize() will
+//  do the initialization, and only the last XMLPlatformUtils::Terminate()
+//  will clean up the memory. The other calls are ignored.
+//
+//  To ensure all the memory held by the parser are freed, the number of
+//  XMLPlatformUtils::Terminate() calls should match the number
+//  of XMLPlatformUtils::Initialize() calls.
+//  -------------------------------------------------------------------
+    void ReInitXerces(void) throw()
```
Comment (tst-lsavoie) [inline earth_enterprise/src/common/khxml/khxml.cpp]: I'd like to see the code for reinitializing xerces also moved to a separate WIP PR. I can see the point of keeping it around, but when we have code that is experimental and is known to cause segfaults, we shouldn't check it in to master.

File: earth_enterprise/src/fusion/autoingest/sysman/khSystemManager.cpp [hunk 13]
```diff
@@ -83,6 +84,9 @@ void systemrc_reload(int signum)
   notify(NFY_WARN, "system log level changed to: %s",
           khNotifyLevelToString(static_cast<khNotifyLevel>(logLevel)).c_str());
   setNotifyLevel(static_cast<khNotifyLevel>(logLevel));
+  // triggering XMLPlatformUtils::Terminate() causes intermittent segfailts
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/khSystemManager.cpp]: Can be removed or moved to a separate PR.

File: earth_enterprise/src/fusion/autoingest/geAssetRoot.cpp [hunk 14]
```diff
@@ -19,7 +19,7 @@
 
 namespace geAssetRoot {
 
-std::string VolumeName = "assetroot";
+const char VolumeName[] = "assetroot";
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/geAssetRoot.cpp]: Minor: it's cleaner to put the memory fixes in a different PR (that applies to this and the other memory fixes included in this PR). I'm not sure if it's worth the effort to pull them out at this point, but it's something to keep in mind for the future. It's best to keep each PR focused on a single purpose.

File: earth_enterprise/src/fusion/autoingest/sysman/khSystemManager.cpp (hunk 13 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/sysman/khSystemManager.cpp]: moving over to the wrapper class PR

**Comment by tst-lsavoie** [comment]
Not sure if you're working on this, but I'm getting compile errors ATM.

**Comment by tst-basilhuffman** [comment]
@tst-lsavoie latest push addresses this

File: earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp [hunk 15]
```diff
@@ -408,14 +415,14 @@ khResourceManager::TryActivate(void) throw()
       taskWaitingQueue.size() &&
       providers.size()) {
     // figure out which providers have CPU resources to spare
-    Providers availProviders;
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: I don't think this line should be deleted. I'm getting a compile error that availProviders was not declared.

Comment (tst-lsavoie) [reply earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: As a separate issue, travis is not catching the error. The build fails but travis marks it as a success.

File: earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp [hunk 16]
```diff
@@ -243,9 +243,13 @@ khResourceManager::InsertProvider(khResourceProviderProxy *proxy)
   // instantiate the Volumes that this provider will manage
   std::vector<std::string> volnames;
   GetHostVolumes(host, volnames);
-  for (std::vector<std::string>::const_iterator vn = volnames.begin();
-       vn != volnames.end(); ++vn) {
-    volumes[*vn] = new Volume(*vn, proxy);
+  for (const auto& vn : volnames) {
+        if (volumes.find(vn) != volumes.end())
+        {
+            delete volumes[vn];
```
Comment (tst-lsavoie) [inline earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: If we're going to include this change in this PR we need a better understanding of what it does. Why do duplicate volumes get added? Is there any difference between the new and old versions of the volumes? I think @tst-jkent is already working on this.

Comment (tst-jkent) [reply earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: I'm still running some tests to make sure, but so far this only gets called when gesystemmanager is started and when systemmanager is stopped the volumes are removed as expected. Still looking into whether or not there are any differences between the volumes each time the volumes are added.

File: earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp (hunk 15 above)
Comment (tst-basilhuffman) [reply earth_enterprise/src/fusion/autoingest/sysman/khResourceManager.cpp]: Has this happened elsewhere vis a vis Travis CI giving false positives?

**Comment by tst-basilhuffman** [comment]
@tst-lsavoie I think the issue vis a vis CLA was a push I did from a unconfiged system. Is there any way to fix this?

**Comment by tst-basilhuffman** [comment]
@tst-lsavoie rebasing looks to not have any effect, it looks to have added a number of additional commits. any other ideas?
